include/kodlab_mjbots_sdk/robot_interface.h

```diff
@@ -0,0 +1,143 @@
+        RobotInterface( std::vector<std::shared_ptr<JointDerived>> joint_vect, 
+                        float robot_max_torque,
+                        int soft_start_duration )
+                        : soft_start_(robot_max_torque, soft_start_duration)
+        {
+            // Ensure at compile time that the template is JointBase or a child of JointBase
+            static_assert(std::is_base_of<JointBase, JointDerived>::value);
+
+            // Set up joint vector and joint states
+            for (std::shared_ptr<JointDerived> j : joint_vect){
+                joints.push_back(j); // Copy constructed, uses an implicit upcasting
+                positions_.push_back( j->get_position_reference() );
+                velocities_.push_back( j->get_velocity_reference() );
+                torque_cmd_.push_back( j->get_servo_torque_reference() ); 
+            }
+
+            // Set number of joints
+            num_joints_ = joint_vect.size();
+        }
+
+        /*!
+         * @brief Destroy the Robot Interface object. Virtual destructor for proper derived pointer destruction.
+         */
+        virtual ~RobotInterface() {};
+
+        /*!
+         * @brief Initialize the robot (e.g. communication or states etc.)
+         */
+        virtual void Init(){};
+
+        /*!
+         * @brief Update the robot state and torque command. 
+         */
+        virtual void Update(){cycle_count_++;}; //TODO remove cycle_count and use time or more intelligently set cycle_count or properly handle softstart
+
+        /*!
+         * @brief Stop the robot by setting torques to zero. Can be overridden
+         *        if other features are available e.g. mehanical or regen braking
+         */
+        virtual void Stop(){SetTorques(std::vector<float>(num_joints_, 0));}
+
+        /*!
+         * @brief accessor for joint positions, takes into account direction and offset
+         * @return the joint positions
+         */
+        std::vector<float> GetJointPositions();
+
+        /*!
+         * @brief accessor for joint velocities, takes into account direction
+         * @return the joint velocities
+         */
+        std::vector<float> GetJointVelocities();
+
+        /*!
+         * @brief setter for torque command
+         * @param torques[in] vector of torques
+         */
+        void SetTorques(std::vector<float> torques);
+
+        /*!
+         * @brief accessor for the torque md, takes into account direction
+         * @return the torque cmd
+         */
+        std::vector<float> GetJointTorqueCmd();
+
+        //TODO virtual Attitude GetAttitude()
+
+        /*! 
+         * @brief Get sub-vector of shared_ptr to joint objects via a set of indices
+         * @param joint_indices set of desired joint indices as std::vector of ints
+         * @return a vector shared pointers to the desired joints
+         */
+        std::vector<std::shared_ptr<JointBase>> GetJoints(std::vector<int> joint_indices);
```

Comment:
Index here by joint_ids rather than by ints.

Reply:
I think this should be handled by maps/typedef as well. In particular, so that we can get the joints by CAN_ID or INDEX with INDEX being of type Joint_Index_t or something like that. Related to #28.